fix: Correctly handle proxy environment variables and arguments

semidark opened the pull request:
## TLDR

This pull request fixes a critical issue where the NO_PROXY environment variable was ignored for LLM requests, preventing users in corporate environments from connecting to internal LLM servers. It also resolves a bug where --proxy='' failed to disable the proxy. The fix ensures that all proxy-related environment variables (HTTP_PROXY, HTTPS_PROXY, NO_PROXY) are respected, and the --proxy command-line argument works as expected.

## Dive Deeper

### Root Causes

1. **NO_PROXY Ignored**:  
   The `GeminiClient` was using Undici’s `ProxyAgent` for LLM requests, but `ProxyAgent` does not support the `NO_PROXY` environment variable. As a result, requests to hosts listed in `NO_PROXY` were still routed through the proxy.

2. **`--proxy=''` Bug**:  
   The configuration logic interpreted an empty string from the `--proxy` flag as falsy, causing the system to fall back to proxy environment variables instead of disabling the proxy entirely.

---

### What This PR Changes

- **Proxy Agent Replacement**:  
  Replaces `ProxyAgent` with `EnvHttpProxyAgent` in the `GeminiClient`. `EnvHttpProxyAgent` correctly reads and respects standard proxy environment variables (`HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY`), aligning LLM client behavior with the IDE client.

- **Improved Proxy Configuration Logic**:  
  The configuration now correctly interprets `--proxy=''` as an explicit instruction to disable the proxy, rather than falling back to environment variables.

- **Hybrid Handling of Proxy Settings**:  
  The `setupProxyConfiguration` method in `GeminiClient` has been refactored to intelligently handle both the `--proxy` argument and environment variables, ensuring correct precedence (i.e., the `--proxy` flag overrides environment variables).

---

### Outcome

These changes result in a robust and predictable proxy implementation that:

- Follows industry proxy standards,
- Honors environment variables (`HTTP_PROXY`, `HTTPS_PROXY`, `NO_PROXY`),
- Respects explicit command-line overrides,
- And behaves as users in corporate environments expect.

---

## Reviewer Test Plan

To validate these changes, replicate the following steps:

1. **Simulate a corporate proxy environment**:
   ```sh
   export HTTP_PROXY="http://your-corporate-proxy:8080"
   export HTTPS_PROXY="http://your-corporate-proxy:8080"
   ```
2. **Exclude internal LLM server from proxy**:
   ```sh
   export NO_PROXY="your-internal-llm-host.com,localhost,127.0.0.1"
   ```
3. **Run `qwen` for internal LLM** (using `OPENAI_*` variables):
   ```sh
   qwen "your prompt"
   ```
   - *Validation*: The command should succeed without connection errors; requests to `your-internal-llm-host.com` should bypass the proxy.

4. **Test disabling proxy with `--proxy=''`**:
   ```sh
   qwen --proxy='' "your prompt"
   ```
   - *Validation*: Command should succeed; the proxy should be disabled for the request.

5. **Test overriding proxy with `--proxy`**:
   ```sh
   qwen --proxy='http://a-different-proxy:8888' "your prompt"
   ```
   - *Validation*: The request should be routed via `http://a-different-proxy:8888` instead of the environment-provided proxy.

---

This reformatting aims to clearly break down the root causes, the implemented fixes, and the outcomes, with explicit steps for functional verification.

## Testing Matrix

<!-- Before submitting please validate your changes on as many of these options as possible -->

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ❓  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Resolves #756

yiliang114 commented:
Regarding --proxy without a parameter: does it typically mean enabling the proxy or not enabling it? It seems there isn’t a very standard convention. However, in the PR it’s used to indicate disabling the proxy. Is there any source or precedent for this? I mean, are there other npm packages or CLIs designed this way? I’d like to use them as a reference.

wenshao commented on packages/core/src/core/client.ts:
```diff
@@ -130,15 +130,57 @@ export class GeminiClient {
+    );
+
+    // If no proxy is configured via --proxy or environment variables, nothing to do
+    if (!configProxy && !hasProxyEnvVars) {
```
**[Critical]** `--proxy=''` is collapsed to `undefined`, so `GeminiClient` cannot distinguish an explicit proxy disable from no CLI proxy override. If `HTTP_PROXY`/`HTTPS_PROXY` is set, `setupProxyConfiguration()` still installs `EnvHttpProxyAgent` from the environment, so the advertised `qwen --proxy=''` behavior continues routing LLM requests through the corporate proxy.

Preserve an explicit-disable state across the config boundary, then skip installing an env proxy dispatcher when that state is present.

_— gpt-5.5 via Qwen Code /review_

wenshao commented on packages/core/src/core/client.ts:
```diff
@@ -130,15 +130,57 @@ export class GeminiClient {
+
+    // If --proxy was specified but no environment variables are set,
+    // temporarily set environment variables for EnvHttpProxyAgent to use
+    if (configProxy && !hasProxyEnvVars) {
```
**[Critical]** A CLI `--proxy` value does not override existing proxy environment variables. When any proxy env var exists, this branch does not pass `configProxy` to `EnvHttpProxyAgent`, so the agent keeps using `HTTP_PROXY`/`HTTPS_PROXY` instead of the CLI-provided proxy.

Construct `EnvHttpProxyAgent` with explicit `httpProxy`/`httpsProxy` values when `configProxy` is present; only fall back to env reading when no CLI override exists.

_— gpt-5.5 via Qwen Code /review_

wenshao commented on packages/core/src/core/client.test.ts:
```diff
@@ -2670,4 +2670,180 @@ ${JSON.stringify(
+      mockConfigObject.getProxy.mockReturnValue(undefined);
+
+      // Act: Create a new client (which calls setupProxyConfiguration)
+      const testClient = new GeminiClient(
```
**[Critical]** The new proxy tests assign `const testClient = new GeminiClient(...)` but never use `testClient`. This occurs in each added proxy test and causes `npm run lint` to fail with `@typescript-eslint/no-unused-vars`; TypeScript also reports `TS6133`.

Instantiate without assignment, or assert on the constructed instance.

_— gpt-5.5 via Qwen Code /review_

wenshao commented on packages/core/src/core/client.ts:
```diff
@@ -130,15 +130,57 @@ export class GeminiClient {
+    // If --proxy was specified but no environment variables are set,
+    // temporarily set environment variables for EnvHttpProxyAgent to use
+    if (configProxy && !hasProxyEnvVars) {
+      // Set both HTTP and HTTPS proxy environment variables
```
**[Suggestion]** `setupProxyConfiguration()` mutates `process.env.HTTP_PROXY` and `process.env.HTTPS_PROXY` when `--proxy` is provided and no env proxy vars exist. That leaks a CLI option into process-global environment state for later code, child processes, telemetry, or unrelated libraries.

Avoid writing to `process.env`; pass explicit proxy options into `EnvHttpProxyAgent` instead.

_— gpt-5.5 via Qwen Code /review_

Commit:
fix(core): respect NO_PROXY env var and --proxy='' for LLM requests

Replace ProxyAgent with EnvHttpProxyAgent so that NO_PROXY is honored
for LLM traffic, matching the existing IDE client behavior. When the CLI
--proxy flag provides a URL it is passed explicitly to override env vars;
--proxy='' disables the proxy entirely, even when HTTP_PROXY/HTTPS_PROXY
are set. The proxy dispatcher is never installed by mutating process.env.

Resolves #756

semidark commented:
## PR Reworked — All Review Comments Addressed

I've rebased this branch onto the latest `main` and re-implemented the fix from scratch to address all four review comments. Here's what changed:

### Review Feedback → Resolution

| # | Review Comment | Resolution |
|---|----------------|------------|
| **Critical #1** — `--proxy=''` collapsed to `undefined`, so `setupProxyConfiguration()` can't distinguish "disable proxy" from "no CLI override" | The CLI layer now returns `false` (not `undefined`) for `--proxy=''`. The `Config.proxy` field type is `string \| false \| undefined`. `setupProxyConfiguration()` returns early when `proxy === false`, skipping proxy setup even when env vars are set. The public `getProxy()` API stays `string \| undefined` — it maps `false` → `undefined` so downstream consumers (OpenAI SDK, Anthropic SDK, telemetry) are unaffected. |
| **Critical #2** — CLI `--proxy` value doesn't override env vars | When `--proxy` provides a URL, it's now passed explicitly as `httpProxy` / `httpsProxy` to `EnvHttpProxyAgent`, which overrides whatever `HTTP_PROXY` / `HTTPS_PROXY` env vars are set. |
| **Critical #3** — Unused `testClient` variables cause lint failures | Tests were rewritten from scratch — no unused variables. |
| **Suggestion #4** — `setupProxyConfiguration()` mutates `process.env` | The method no longer writes to `process.env`. Proxy options are passed directly to `EnvHttpProxyAgent` constructor. |

### Architecture Change

The proxy setup was moved from `GeminiClient` into `Config` (where it already lives on upstream `main`). The key change is replacing `ProxyAgent` with `EnvHttpProxyAgent`, which natively supports `NO_PROXY`. The proxy dispatcher is configured in `Config`'s constructor via a private `setupProxyConfiguration()` method.

### Data Flow

```
CLI --proxy flag
    │
    ▼
getProxyConfiguration(argv.proxy)
    --proxy="url"  → "url"
    --proxy=""     → false  (explicit disable)
    not provided   → HTTPS_PROXY > https_proxy > HTTP_PROXY > http_proxy
    │
    ▼
Config.proxy: string | false | undefined
    │
    ▼
Config.setupProxyConfiguration()
    false          → return (no proxy, even with env vars)
    undefined + no env vars → return (nothing to do)
    string         → EnvHttpProxyAgent({ httpProxy, httpsProxy, noProxy })
    undefined + env vars   → EnvHttpProxyAgent({ noProxy })  // reads from env
```

### Tests Added

**Core (`packages/core/src/config/config.test.ts`)** — 7 new tests:
- No proxy dispatcher when nothing configured
- `EnvHttpProxyAgent` uses env vars when only env vars are set
- Explicit `httpProxy`/`httpsProxy` passed when `--proxy` is provided
- CLI `--proxy` overrides env vars
- No dispatcher when proxy explicitly disabled (`false`)
- `NO_PROXY` values preserved + localhost variants added
- `process.env` never mutated

**CLI (`packages/cli/src/config/config.test.ts`)** — 3 new tests:
- `--proxy=""` disables proxy
- Fallback to env vars when `--proxy` not provided
- Env var precedence (`HTTPS_PROXY` > `https_proxy` > `HTTP_PROXY` > `http_proxy`)

### Verification

- ✅ Build: clean
- ✅ Typecheck: clean
- ✅ Lint: clean
- ✅ Core tests: 6056 passed (245 files)
- ✅ CLI tests: 4445 passed (282 files)

wenshao commented on packages/core/src/config/config.ts:
```diff
@@ -2032,9 +2029,67 @@ export class Config {
+      return;
+    }
+
+    const configProxy = this.getProxy();
```
**[Critical]** This only makes the global dispatcher `NO_PROXY`-aware. The OpenAI-compatible, DashScope, and Anthropic clients still call `buildRuntimeFetchOptions(..., this.cliConfig.getProxy())`, and that helper constructs its own `ProxyAgent` dispatcher. That per-client dispatcher overrides the global `EnvHttpProxyAgent`, so model SDK traffic will still ignore `NO_PROXY` whenever a configured proxy is present. Please make the runtime fetch options use the same `NO_PROXY`-aware behavior, or rely on the global dispatcher consistently while preserving the timeout settings.

_— gpt-5.5 via Qwen Code /review_

semidark replied:
Good catch — fixed in fc7def79c.

The per-client `ProxyAgent` in `buildFetchOptionsWithDispatcher()` was indeed overriding the global `EnvHttpProxyAgent`, so SDK traffic (OpenAI, DashScope, Anthropic) was still ignoring `NO_PROXY`.

Changes:

1. **`runtimeFetchOptions.ts`** — Replaced `ProxyAgent` with `EnvHttpProxyAgent` in `buildFetchOptionsWithDispatcher()`. When a proxy URL is present, it's passed via `httpProxy`/`httpsProxy` with `noProxy` from `buildNoProxyList()`, plus the same `headersTimeout: 0` / `bodyTimeout: 0` timeout disabling. When no proxy URL is given, still uses plain `Agent` with disabled timeouts.

2. **`proxyUtils.ts`** — Extracted `buildNoProxyList()` helper to keep the `NO_PROXY` list construction DRY between `Config.setupProxyConfiguration()` and the per-client dispatchers. Both now call the same function.

3. **Tests updated** — `runtimeFetchOptions.test.ts` asserts `httpProxy`, `httpsProxy`, `noProxy` on the dispatcher. `proxyUtils.test.ts` has 4 new tests for `buildNoProxyList()`.

There are no more `ProxyAgent` imports in production code — all proxy dispatchers now use `EnvHttpProxyAgent`.

Commit:
fix(core): use EnvHttpProxyAgent in SDK clients so NO_PROXY is respected

The OpenAI, DashScope, and Anthropic SDK clients were constructing their
own ProxyAgent via buildRuntimeFetchOptions(), which overrode the global
EnvHttpProxyAgent and ignored NO_PROXY. Replace ProxyAgent with
EnvHttpProxyAgent in buildFetchOptionsWithDispatcher() and extract a
shared buildNoProxyList() helper into proxyUtils.ts to keep the logic
DRY between Config.setupProxyConfiguration() and the per-client
dispatchers.

semidark committed:
Merge branch 'QwenLM:main' into fix-756

wenshao commented on packages/cli/src/config/config.ts:
```diff
@@ -1269,3 +1264,31 @@ export async function loadCliConfig(
+  }
+
+  // Fall back to environment variables in precedence order
+  return (
```
**[Critical]** `getProxyConfiguration()` resolves uppercase proxy environment variables before lowercase, but Undici's `EnvHttpProxyAgent` resolves lowercase first (`https_proxy` before `HTTPS_PROXY`, `http_proxy` before `HTTP_PROXY`). If both cases are set differently, `Config.getProxy()` and SDK runtime fetch options can use one proxy while the global Undici dispatcher uses another, making proxy behavior inconsistent.

```suggestion
    process.env['https_proxy'] ||
    process.env['HTTPS_PROXY'] ||
    process.env['http_proxy'] ||
    process.env['HTTP_PROXY']
```

_— gpt-5.5 via Qwen Code /review_

wenshao commented on packages/core/src/utils/proxyUtils.ts:
```diff
@@ -20,6 +20,22 @@
  * @returns The normalized proxy URL with protocol prefix, or undefined if input is undefined/empty
  * @throws Error if a SOCKS proxy URL is provided
```
**[Suggestion]** The JSDoc for `normalizeProxyUrl()` is now detached because `buildNoProxyList()` was inserted between the comment and the function. This can make editor hover/generated docs associate the normalization documentation with the wrong symbol.

Move the `normalizeProxyUrl()` JSDoc so it appears immediately above `normalizeProxyUrl()`, after `buildNoProxyList()`.

_— gpt-5.5 via Qwen Code /review_
